test/test_temperature.py

```diff
@@ -3,22 +3,127 @@
 import numpy as np
 import pytest
 import ufl
+from ufl.conditional import Conditional
 
 test_mesh = F.Mesh1D(vertices=np.array([0.0, 1.0, 2.0, 3.0, 4.0]))
 x = ufl.SpatialCoordinate(test_mesh.mesh)
+dummy_mat = F.Material(D_0=1, E_D=1, name="dummy_mat")
 
 
 @pytest.mark.parametrize(
-    "value", [1, fem.Constant(test_mesh.mesh, 1.0), 1.0, "coucou", 2 * x[0]]
+    "value", [1, fem.Constant(test_mesh.mesh, 1.0), 1.0, "coucou", lambda x: 2 * x[0]]
 )
 def test_temperature_type_and_processing(value):
     """Test that the temperature type is correctly set"""
-    my_model = F.HydrogenTransportProblem()
-    my_model.mesh = test_mesh
+    my_model = F.HydrogenTransportProblem(
+        mesh=F.Mesh1D(vertices=np.linspace(0.0, 1.0)),
+        subdomains=[
+            F.VolumeSubdomain1D(1, borders=[0, 1], material=F.Material(1, 1, "my_mat"))
+        ],
+        species=[F.Species("H")],
+    )
+    my_model.temperature = value
+    my_model.settings = F.Settings(atol=1, rtol=0.1, final_time=2)
+    my_model.settings.stepsize = F.Stepsize(initial_value=1)
 
     if not isinstance(value, (fem.Constant, int, float)):
-        with pytest.raises(TypeError):
-            my_model.temperature = value
+        if callable(value):
+            my_model.initialise()
+        else:
+            with pytest.raises(TypeError):
+                my_model.initialise()
     else:
-        my_model.temperature = value
-        assert isinstance(my_model.temperature, fem.Constant)
+        my_model.initialise()
+        assert isinstance(my_model.temperature_fenics_value, fem.Constant)
```

Comment:
I see there's quite the logic in this test. Let's minimise it by splitting the test into two: one that tests against valid inputs, another that tests against invalid inputs

Reply:
Also, related to another comment I had above, if we keep the setter for `HydrogenTransportProblem.temperature`, then I don't think we need to modify all this